### PR TITLE
Refactor plugin reconciliation for dependency mechanism

### LINKED ISSUE #5872: 建议重构插件依赖相关的插件启动逻辑
### Your current Halo version

2.15.1

### Describe this feature

**目前插件依赖中关于被依赖插件的停止或启动存在一些问题：**
1. 以应用市场插件为例，应用市场插件被 A 和 B 插件依赖，且都处于启动状态
2. 停止应用市场插件后会自动停止 A 和 B，然后点击启动 A 会自动启动 B 插件和应用市场插件
3. 在 Reconciler 中会去监听内存中插件的实际状态并修改 spec.enabled 这违反了 spec 的初衷，spec 是用户期望如果内存中状态不符合应该协调到期望状态

**重构点：**
1. 插件的停止和启动前应该加入依赖检测，如果有插件依赖了当前被停止的插件则不允许停止当前插件，如果插件启动时所依赖的插件没有启动应该拒绝启动并给用户提示
2. 不应该监听内存中插件状态并去修改 spec 来适应内存中的状态，而是要以用户期望状态为准

**以升级步骤为例，应该的逻辑是：**
1. 检测依赖插件，比如已经启动的 A，B
2. 更新 A 和 B 的 annotations 中加入 reload 标记
3. 检测依赖插件，有就抛出异常 requeue（可能因为抛出异常 requeue 后进入第4步），没有则 reload 应用市场，stop ,unload,load 等步骤
4. A 和 B 中途可能会进入执行（由于 3 步骤让出了执行权），检测到 reload 标记，执行 reload 操作，但是需要检测依赖插件是否正常，如果没有启动则抛出异常等待。
5. 应用市场启动完成了，reconciler 轮到 A 执行，检测到 reload 标记执行 reload 等逻辑重复到第3步

如此不在需要停止插件时检测依赖并记录到当前插件然后又再启动后恢复依赖插件的状态，且不需要去以内存中实际状态为主来修改插件 spec.enabled，spec.enabled 是用户期望,无论如何要以用户期望为准，避免出现升级应用市场插件时其他插件被停止而没有自动启动的问题

### Additional information

/kind improvement
/area core
/milestone 2.16.x
/assign @JohnNiang

### PR DESCRIPTION
**Pull request by JohnNiang**
#### What type of PR is this?

/kind improvement
/area core
/area plugin

#### What this PR does / why we need it:

This PR wholly refactors plugin reconciliation to implement dependency mechanism.

Currently,
- If we disable plugin which has dependents, the plugin must wait for dependents to be disabled.
- If we enable plugin which has dependencies , the plugin must wait for dependencies to be enabled.
- If we upgrade plugin which has dependents, the plugin must request dependents to be unloaded. After the plugin is unloaded, the plugin must cancel unload request for dependents.

#### Which issue(s) this PR fixes:

Fixes #5872 

#### Does this PR introduce a user-facing change?

```release-note
优化被依赖的插件的升级，启用和禁用
```

### CODE REVIEW

File: application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java [hunk 1]
```diff
@@ -77,8 +75,9 @@
 @Component
 public class PluginReconciler implements Reconciler<Request> {
     private static final String FINALIZER_NAME = "plugin-protection";
-    private static final String DEPENDENTS_ANNO_KEY = "plugin.halo.run/dependents-snapshot";
```
Comment (guqing) [inline application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java:80]: 直接删除这里是否存在考虑历史数据残留问题

File: application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java [hunk 2]
```diff
@@ -386,44 +437,121 @@ private void resolveStaticResources(Plugin plugin) {
...
+            .status(ConditionStatus.TRUE)
+            .reason(ConditionReason.LOADED)
+            .lastTransitionTime(clock.instant())
+            .build());
+        plugin.getStatus().setPhase(Plugin.Phase.RESOLVED);
+        return null;
     }
 
-    private void createOrUpdateSetting(Plugin plugin) {
+    private Result createOrUpdateSetting(Plugin plugin) {
```
Comment (guqing) [inline application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java:548]: 这个方法始终返回 null 值，那么可以不需要返回值，直接在  steps.add 处处理

File: application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java (hunk 1 above)
Comment (JohnNiang) [reply application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java:80]: 可以考虑在 Reconciler 中直接移除掉这个 Annotation。

File: application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java (hunk 2 above)
Comment (JohnNiang) [reply application/src/main/java/run/halo/app/core/extension/reconciler/PluginReconciler.java:548]: 1. 建议保留同一个风格。
2. 后期在该方法内有可能会尝试返回不同的 result。

**Comment by guqing** [comment]
建议为 plugin-state API 加一个前置的依赖检查，如果有依赖没有启动直接报错否则前端会一直 pending 而用户不知道为什么

**Review by ruibaby — COMMENTED** [review]
/lgtm

**Review by guqing — APPROVED** [review]
/approve

**Pull request merged**